File: Programmers/Lv4/Lv4_징검다리.py
# ?뚮씪硫뷀듃由??쒖튂
def remove_rock(rocks, mid, n):
    prev = 0
    remove = 0

    for r in rocks:
        if r - prev > mid:
            prev = r
        else:       # mid 蹂대떎 ?묒? 媛믪씠 ?덈떎硫??쒓굅?댁빞 ????媛?닔 利앷?
            remove += 1
    if remove > n:  # ?뺥빐吏??뚮낫??留롮씠 ?쒓굅?댁빞 ?쒕떎硫?
        return True
    else:
        return False


def solution(distance, rocks, n):
    low = 0
    high = distance
    rocks.sort()
    while low <= high:
        mid = (low + high) >> 1         # mid => ?뺣떟媛??덉륫
        if remove_rock(rocks, mid, n):
            high = mid - 1              # mid 媛믪씠 ?믪? 寃?
        else:
            low = mid + 1               # mid 媛믪씠 ??? 寃?
    return low

print(solution(25, [2, 14, 11, 21, 17], 2))

File: Programmers/Lv4/Lv4_징검다리.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

bool remove_rock(vector<int> &rocks, int min, int n) {
	int prev = 0;
	int remove = 0;
	for (int r : rocks) {
		if (r - prev > min)
			prev = r;
		else
			remove++;
	}
	if (remove > n)
		return true;
	else
		return false;
}

// 파라메트릭 서치 이용!
int solution(int distance, vector<int> rocks, int n) {
	sort(rocks.begin(), rocks.end());
	int low = 0;
	int mid = 0;
	int high = distance;

	while (low <= high) {
		mid = (low + high) / 2;	
		if (remove_rock(rocks, mid, n))		// 돌을 n보다 많이 제거 했다면
			high = mid - 1;					// high를 줄이고
		else								// 돌을 n보다 적게 제거 했다면
			low = mid + 1;					// low를 올리자
	}
	return low;
}

